### PR TITLE
gh-119793: Add optional length-checking to `map()`

### PR DESCRIPTION
**Pull request by nineteendo**
> These two examples silently truncate the unmatched inputs:
> 
> ```python
> >>> list(map(pow, [1, 2, 3], [2, 2, 2, 2]))
> [1, 4, 9]
> >>> list(map(pow, [1, 2, 3, 4, 5], [2, 2, 2, 2]))
> [1, 4, 9, 16]
> ```

* Issue: gh-119793

----
📚 Documentation preview 📚: [`library/functions.html#map`](https://cpython-previews--120471.org.readthedocs.build/en/120471/library/functions.html#map)

### CODE REVIEW

File: Lib/test/test_itertools.py
Comment (nineteendo) [inline Lib/test/test_itertools.py:2424]: Is there a clean way to support this? Because this worked previously.

File: Lib/test/test_builtin.py
Comment (nineteendo) [inline Lib/test/test_builtin.py:1257]: How am I supposed to test this? Pickle can't save lambda's.

File: Lib/test/test_builtin.py
Comment (nineteendo) [inline Lib/test/test_builtin.py]: How am I supposed to test this? `pickle` can't save lambda's. And I wanted to keep the tests as much in line with `zip()` as possible.

File: Python/bltinmodule.c
Comment (eendebakpt) [inline Python/bltinmodule.c]: Why is the vectorcall variant removed?

Comment (nineteendo) [reply Python/bltinmodule.c]: `map_new()` seemed to be never called because of `map_vectorcall()`.
And the repeated code screamed refactoring to me, but I'll revert it as it's apparently faster.

File: Python/bltinmodule.c
Comment (nineteendo) [inline Python/bltinmodule.c:1385]: This feels very ugly, copied from `unicode_new()`. Should I do this differently?

File: Doc/library/functions.rst
Comment (picnixz) [inline Doc/library/functions.rst]: Why do we have a `\.` at then end?

File: Doc/library/functions.rst
Comment (picnixz) [inline Doc/library/functions.rst]:
```suggestion
      Added the *strict* argument.
```

For parameter names, we don't use literals but rather emphasis.

File: Lib/test/test_builtin.py
Comment (picnixz) [inline Lib/test/test_builtin.py:1277]: Just wondering but are those tests based on the `zip` tests? if so, that's fine.

File: Python/bltinmodule.c
Comment (picnixz) [inline Python/bltinmodule.c:1431]: Can you just add a comment to know that *i* is declared here for error reporting afterwards?

File: Python/bltinmodule.c
Comment (picnixz) [inline Python/bltinmodule.c]:
```suggestion
        PyObject *ret = map_new(tp, tuple, dict);
```

File: Lib/test/test_builtin.py
Comment (nineteendo) [reply Lib/test/test_builtin.py:1277]: Yes, I was too lazy to write new tests. I've added a comment.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1431]:
```suggestion
    Py_ssize_t i; // Needed for error reporting
```

File: Doc/library/functions.rst
Comment (nineteendo) [reply Doc/library/functions.rst]: Let's try without:

```suggestion
   see :func:`itertools.starmap`.
```

File: Python/bltinmodule.c
Comment (nineteendo) [inline Python/bltinmodule.c]: This feels very ugly, copied from `unicode_new()`. Should I do this differently?

**Comment by nineteendo** [comment]
@rhettinger, thoughts? You requested this.

File: Doc/whatsnew/3.14.rst
Comment (nineteendo) [inline Doc/whatsnew/3.14.rst]: Is this the correct section to put this?

File: Doc/library/functions.rst
Comment (picnixz) [inline Doc/library/functions.rst]: Strictly speaking, *strict* is a parameter and not an argument (we do however refer to its value as an argument but I think changes in signatures use the wording "parameters" (which is formally correct)).

File: Doc/whatsnew/3.14.rst
Comment (picnixz) [reply Doc/whatsnew/3.14.rst]: Maybe check where it has been put when `strict` has been added for `zip`.

File: Doc/library/functions.rst
Comment (nineteendo) [reply Doc/library/functions.rst]:
```suggestion
      Added the *strict* parameter.
```

Comment (nineteendo) [reply Doc/library/functions.rst]: > Strictly

Pun intended?

Comment (picnixz) [reply Doc/library/functions.rst]: Not all :rofl: !

File: Doc/whatsnew/3.14.rst
Comment (nineteendo) [reply Doc/whatsnew/3.14.rst]: Yes, I correctly guessed it: https://docs.python.org/3.14/whatsnew/3.10.html#other-language-changes
I copied the sentence from there.

File: Doc/library/functions.rst
Comment (nineteendo) [inline Doc/library/functions.rst]: Should we add examples like for [`zip()`](https://docs.python.org/dev/library/functions.html#zip)? It's also worth noting `itertools.map_longest()` doesn't exist.

**Comment by nineteendo** [comment]
cc @brandtbucher

File: Doc/library/functions.rst
Comment (encukou) [reply Doc/library/functions.rst]: I don't think that's necessary.

File: Doc/whatsnew/3.14.rst
Comment (encukou) [inline Doc/whatsnew/3.14.rst]: I'd rather see ”like `zip` than a reference to PEP-618, which lists this in Rejected ideas.

File: Lib/test/test_itertools.py
Comment (encukou) [reply Lib/test/test_itertools.py:2424]: Overriding `__new__`?
```py
                class subclass_with_init(cls):
                    def __init__(self, *args, newarg=None):
                        self.newarg = newarg
                    if cls in [repeat, zip, map, compress]:
                        # Constructors of repeat, zip, map, compress accept keyword arguments.
                        # Their subclasses need an overridden __new__ that ignores
                        # the new keyword arguments.
                        def __new__(cls, *args, newarg=None, **kwargs):
                            return super().__new__(cls, *args, **kwargs)
```

File: Python/bltinmodule.c
Comment (encukou) [reply Python/bltinmodule.c]: The nice-looking way is:
```c
        // Fallback to map_new()
        PyThreadState *tstate = _PyThreadState_GET();
        return _PyObject_MakeTpCall(tstate, type, args, nargsf, kwnames);
```
But I'm not sure about performance.

File: Python/bltinmodule.c
Comment (encukou) [inline Python/bltinmodule.c:1483]: At a first glance, 1-2 looks like “one minus two”.
Would “previous argument[s]” would work better?

File: Python/bltinmodule.c
Comment (encukou) [inline Python/bltinmodule.c:1537]: Can this be a tuple in the future? Maybe compare to actual Py_True.

**Review by encukou — COMMENTED** [review]
I'd wait for the [recent discussion about visibility](https://discuss.python.org/t/62235/) to settle down, before making a change to a builtin.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1483]: This matches the exception of `zip()`.

File: Lib/test/test_itertools.py
Comment (nineteendo) [reply Lib/test/test_itertools.py:2424]: I guess that's acceptable. I wasn't sure if it was fine breaking this.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1537]: This is also what `zip()` does, should I deviate?

**Comment by encukou** [comment]
@rhettinger, do you want to answer questions about [breaking subclasses that add kwargs](https://github.com/python/cpython/pull/120471/files#r1638758992) and [interpreting pickle state](https://github.com/python/cpython/pull/120471/files#r1743830347)?

I can give my answers but this might be too subjective :)

**Comment by encukou** [comment]
To help this feature in, I'll give my answers on these details.

File: Python/bltinmodule.c
Comment (encukou) [reply Python/bltinmodule.c:1537]: I'd prefer if you deviate, to make this a bit more forward-compatible/future-proof.

File: Lib/test/test_itertools.py
Comment (encukou) [reply Lib/test/test_itertools.py:2424]: IMO, it's fine. If it becomes a problem we should fix all of these.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1537]:
```suggestion
    // For now we keep things strict, rather than using PyObject_IsTrue().
    int strict = state == Py_True;
```

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c]: `type` and `int` use it too, so it should be fine.

```suggestion
    if (kwnames != NULL && PyTuple_GET_SIZE(kwnames) != 0) {
        // Fallback to map_new()
        PyThreadState *tstate = _PyThreadState_GET();
        return _PyObject_MakeTpCall(tstate, type, args, nargsf, kwnames);
    }
```

**Comment by nineteendo** [comment]
I've resolved all conversations, I hope this can be merged now.

**Review by encukou — APPROVED** [review]

**Comment by nineteendo** [comment]
Thanks encukou, do you know someone besides Brandt and Reymond that might want to review this? I think it would be best to have more than 1 approval.

**Comment by encukou** [comment]
It would, but if no one finds the time, I plan to merge it next week.

File: Python/bltinmodule.c
Comment (eendebakpt) [inline Python/bltinmodule.c]:
```suggestion
    int strict = state == Py_True;
    lz->strict = strict;
```
This part of the code could never be reached right?

File: Doc/whatsnew/3.14.rst
Comment (erlend-aasland) [inline Doc/whatsnew/3.14.rst]: [Please see the devguide](https://devguide.python.org/documentation/markup/#quick-reference) for how to properly mark up parameters and arguments.

```suggestion
* The :func:`map` function now has an optional *strict* flag like
```

File: Misc/NEWS.d/next/Core_and_Builtins /2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Core_and_Builtins /2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst]:
```suggestion
:func:`map` now supports :pep:`618`'s *strict* parameter, which raises a :exc:`ValueError` if the arguments are exhausted at different lengths.
```

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c:1306]: Consider including `stdbool.h` and making this a `bool`.
`git grep stdbool Python/` for inspiration.

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c]:
```suggestion
    Py_ssize_t i;
```

Moreover, `i` is now shadowed in L1466-1468.

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c:1450]: It's preferred to align modified code with PEP-7.

```suggestion
    for (i = 0; i < niters; i++) {
```

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c:1484]: PEP-7 nit

```suggestion
        const char *plural = i == 1 ? " " : "s 1-";
```

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c]:
```suggestion
    goto exit;
```

File: Python/bltinmodule.c
Comment (erlend-aasland) [inline Python/bltinmodule.c:1476]: This comment explains literally what the code reads. I don't think it adds much value. Ditto for L1501.

File: Doc/whatsnew/3.14.rst
Comment (nineteendo) [reply Doc/whatsnew/3.14.rst]: This sentence was copied from here: https://docs.python.org/3/whatsnew/3.10.html#other-language-changes
Do I change it there too?

**Comment by nineteendo** [comment]
Erlend, your remaining suggestions also apply to zip, as the code was copied. What do you suggest I do?

File: Doc/whatsnew/3.14.rst
Comment (erlend-aasland) [reply Doc/whatsnew/3.14.rst]: No

File: Doc/whatsnew/3.14.rst
Comment (picnixz) [inline Doc/whatsnew/3.14.rst]:
```suggestion
* The :func:`map` built-in now has an optional keyword-only *strict* flag
  like :func:`zip` to check that all the iterables are of equal length.
  (Contributed by Wannes Boeykens in :gh:`119793`.)
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst]:
```suggestion
Patch by Wannes Boeykens.
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-06-13-19-12-49.gh-issue-119793.FDVCDk.rst]: If you update the What's New entry, don't forget to update this one.

File: Python/bltinmodule.c
Comment (picnixz) [reply Python/bltinmodule.c:1306]: Agreed. Currently `zip` converts objects to their truthy/falsey value for `strict` but I think it's better we expect boolean-like values and not just arbitrary objects. What do you think?

File: Python/bltinmodule.c
Comment (picnixz) [reply Python/bltinmodule.c:1431]: I think you forgot to commit this.

File: Python/bltinmodule.c
Comment (picnixz) [inline Python/bltinmodule.c]:
```suggestion
    {"__setstate__", _PyCFunction_CAST(map_setstate), METH_O, setstate_doc},
```

File: Python/bltinmodule.c
Comment (picnixz) [reply Python/bltinmodule.c:1476]: I think this was just C/C from `zip` which  included this kind of comment.

File: Python/bltinmodule.c
Comment (picnixz) [inline Python/bltinmodule.c]: This comment might seem contradictory with the constructor which computes the object truthiness. So I think you should do the same as it's done in `zip` or make the constructor reject non-boolean values.

File: Lib/test/test_builtin.py
Comment (picnixz) [inline Lib/test/test_builtin.py:1376]: The constructor accepts arbitrary object as `strict` so I think we should that it's indeed the case and that pickling/unpickling works properly.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1431]: Erlend suggested to revert this.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c:1476]: Yeah, I was still waiting for a response on what do with it in that case.

File: Python/bltinmodule.c
Comment (nineteendo) [reply Python/bltinmodule.c]: Hmm, let's revert that change then. We would want `zip()` and `map()` to act the same anyway when strict is set.

File: Doc/library/functions.rst
Comment (rhettinger) [inline Doc/library/functions.rst]: `arguments` -> `inputs`

File: Python/bltinmodule.c
Comment (rhettinger) [reply Python/bltinmodule.c:1306]: I see no need to diverge from the equivalent code in `zip` which currently uses `int`:
```
typedef struct {
    PyObject_HEAD
    Py_ssize_t tuplesize;
    PyObject *ittuple;     /* tuple of iterators */
    PyObject *result;
    int strict;
} zipobject;
```

File: Python/bltinmodule.c
Comment (rhettinger) [reply Python/bltinmodule.c:1450]: We use `i=0` in many places including this module.  PEP 7 is not specific this and recommends mostly matching other code in the mode.  In this case, the code is being modeled after `zip()`.

**Review by rhettinger — APPROVED** [review]

File: Python/bltinmodule.c
Comment (erlend-aasland) [reply Python/bltinmodule.c:1431]: If you declare an iteration index variable in the beginning of a scope, it is implicit that it is needed for the entire scope of the function, IMO. Comments should IMO be reserved for the parts of the source code where it is not immediately obvious what is happening or why something was added. An index variable is IMO too trivial to deserve a comment :)

File: Python/bltinmodule.c
Comment (erlend-aasland) [reply Python/bltinmodule.c:1306]: > I see no need to diverge from the equivalent code in zip which currently uses int

We've steadily introduced `stdbool.h` and the `bool` type to the CPython source code; `git grep stdbool` for enlightenment. Using `bool` here results in clearer, more readable, and more maintainable code.

File: Doc/library/functions.rst
Comment (erlend-aasland) [reply Doc/library/functions.rst]: Or perhaps:

```suggestion
   iterables is exhausted before the others, a :exc:`ValueError` is raised. For
```

File: Python/bltinmodule.c
Comment (picnixz) [reply Python/bltinmodule.c:1431]: My rationale was that ANSI C requires to declare variables at the top of the scope. Here it's because we need to remember when we stoped not because of that convention (which is used a lot by old code). (That's why for me it wasn't a "trivial" comment)

File: Lib/test/test_builtin.py
Comment (nineteendo) [reply Lib/test/test_builtin.py:1376]: Only supporting booleans requires more code, so can't we assume the implementation is as simple as possible? Only 11.4k files even use strict zip: [`/zip\(.*strict=/`](https://github.com/search?q=%2Fzip%5C%28.*strict%3D%2F+lang%3APython+&type=code)

Comment (picnixz) [reply Lib/test/test_builtin.py:1376]: Per Raymond's comment, I think we can just keep the same implementation as `zip`. Actually, my comment was about "testing" that arbitrary objects are evaluated properly.

**Review by erlend-aasland — APPROVED** [review]

File: Lib/test/test_builtin.py
Comment (nineteendo) [reply Lib/test/test_builtin.py:1376]: Yeah, I know that's what you meant, but I'm wondering if we ever test that truthy values are accepted in addition to booleans.

This is simpler...

```c
PyArg_ParseTupleAndKeywords(..., "p", &bool)
```

... than only accepting True:

```c
PyArg_ParseTupleAndKeywords(..., "O", &obj)

int bool = obj == Py_True;
```

**Pull request merged**
